File: apps/CounselCompanion/ios/CounselCompanion/LLM/ZeticChatEngine.swift
import Foundation
import ZeticMLange

actor ZeticChatEngine {
    private var model: ZeticMLangeLLMModel?
    private var stopRequested = false

    static let modelId = "Steve/kanana_1_5__2_1b_instruct"
    static let personalKey = "YOUR_MLANGE_KEY"

    var isInitialized: Bool { model != nil }

    init() {}

    func initialize(onDownloadProgress: @escaping @Sendable (Float) -> Void) throws {
        guard model == nil else { return }
        model = try ZeticMLangeLLMModel(personalKey: Self.personalKey, name: Self.modelId) { progress in
            onDownloadProgress(progress)
        }
    }

    func generate(prompt: String, onToken: @Sendable @escaping (String) -> Void) async -> GenerationRecord {
        guard let model = model else {
            return GenerationRecord(text: "", tokenCount: 0, durationMs: 0, stopped: false, error: "Model not initialized")
        }
        try? model.cleanUp()
        stopRequested = false

        let start = DispatchTime.now()
        var text = ""
        var tokenCount = 0
        var err: String?

        do {
            _ = try model.run(prompt)
            while true {
                if stopRequested { break }
                let result = model.waitForNextToken()
                if result.generatedTokens == 0 { break }
                if !result.token.isEmpty {
                    text.append(result.token)
                    tokenCount += result.generatedTokens
                    await MainActor.run { onToken(result.token) }
                }
                try Task.checkCancellation()
            }
        } catch {
            err = error.localizedDescription
        }

        try? model.cleanUp()
        let durationMs = Int((DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000)
        return GenerationRecord(text: text, tokenCount: tokenCount, durationMs: durationMs, stopped: stopRequested, error: err)
    }

    func requestStop() {
        stopRequested = true
        try? model?.cleanUp()
    }

    func destroy() {
        stopRequested = true
        try? model?.cleanUp()
    }

    static func maskedKey() -> String {
        let k = personalKey
        guard k.count >= 8 else { return "****" }
        return k.prefix(4) + "****" + k.suffix(4)
    }
}

File: apps/CounselCompanion/android/app/src/main/java/com/yeonseok/melangecounsel/llm/ZeticChatEngine.kt
package com.yeonseok.melangecounsel.llm

import android.content.Context
import com.zeticai.mlange.core.model.llm.ZeticMLangeLLMModel
import com.zeticai.mlange.core.model.ModelLoadingStatus
import com.yeonseok.melangecounsel.domain.GenerationRecord
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicBoolean

class ZeticChatEngine(private val context: Context) {
    private var model: ZeticMLangeLLMModel? = null
    private val mutex = Mutex()
    private val stopRequested = AtomicBoolean(false)

    val isInitialized: Boolean get() = model != null

    fun initialize(onLoadingStatus: (ModelLoadingStatus) -> Unit) {
        if (model == null) {
            model = ZeticMLangeLLMModel(context.applicationContext, PERSONAL_KEY, MODEL_ID) { status ->
                onLoadingStatus(status)
            }
        }
    }

    suspend fun generate(prompt: String, onToken: suspend (String) -> Unit): GenerationRecord {
        val llm = model ?: return GenerationRecord(
            fullText = "", tokenCount = 0, durationMs = 0,
            stopped = false, errorMessage = "Model not initialized"
        )
        return withContext(Dispatchers.IO) {
            mutex.withLock {
                stopRequested.set(false)
                llm.cleanUp()

                val start = System.currentTimeMillis()
                val completeText = StringBuilder()
                var tokenCount = 0
                var errorMessage: String? = null

                try {
                    llm.run(prompt)
                    coroutineScope {
                        while (isActive && !stopRequested.get()) {
                            ensureActive()
                            val tokenResult = llm.waitForNextToken()
                            if (tokenResult.generatedTokens == 0) break
                            if (tokenResult.token.isNotEmpty()) {
                                completeText.append(tokenResult.token)
                                tokenCount += tokenResult.generatedTokens
                                onToken(tokenResult.token)
                            }
                        }
                    }
                } catch (t: Throwable) {
                    errorMessage = t.message ?: "Generation failed"
                } finally {
                    llm.cleanUp()
                }

                val end = System.currentTimeMillis()
                GenerationRecord(
                    fullText = completeText.toString(),
                    tokenCount = tokenCount,
                    durationMs = end - start,
                    stopped = stopRequested.get(),
                    errorMessage = errorMessage
                )
            }
        }
    }

    fun requestStop() {
        stopRequested.set(true)
        model?.cleanUp()
    }

    fun destroy() {
        stopRequested.set(true)
        model?.cleanUp()
    }

    companion object {
        const val MODEL_ID: String = "Steve/kanana_1_5__2_1b_instruct"
        private const val PERSONAL_KEY: String = "YOUR_MLANGE_KEY"

        fun maskedPersonalKey(): String {
            return if (PERSONAL_KEY.length < 8) {
                "****"
            } else {
                PERSONAL_KEY.take(4) + "****" + PERSONAL_KEY.takeLast(4)
            }
        }
    }
}
